### PR TITLE
[cmd_vel_smoother] add cmd_vel_smoother

### PR DESCRIPTION
**Pull request by furushchev**
This should solve https://github.com/jsk-ros-pkg/jsk_robot/issues/198 https://github.com/jsk-ros-pkg/jsk_common/issues/160
Tested on PR1012

when publishing rate of original `cmd_vel` (blue line) is under control loop, this node publishes interpolating messages (red line)
(monitoring if robot violates acceleration/deceleration (occured by rapid stop))

cmd_vel_smoothed

### CODE REVIEW

File: cmd_vel_smoother/package.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<package>
+  <name>cmd_vel_smoother</name>
+  <version>0.0.0</version>
```
Comment (garaemon) [inline cmd_vel_smoother/package.xml]: version number

File: cmd_vel_smoother/src/cmd_vel_smoother.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  void velCallback(const geometry_msgs::Twist::ConstPtr &msg){
+    latest_time_ = ros::Time::now();
+    latest_vel_ = msg;
+    count_ = 0;
+    pub_.publish(msg);
+  }
+
+  void timerCallback(const ros::TimerEvent& event){
+    loadParams();
```
Comment (garaemon) [inline cmd_vel_smoother/src/cmd_vel_smoother.cpp]: dynamic_reconfigure may be better

File: cmd_vel_smoother/src/cmd_vel_smoother.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    pnh_.param<double>("~x_acc_lim", x_acc_lim_, 0.1);
+    pnh_.param<double>("~y_acc_lim", y_acc_lim_, 0.1);
+    pnh_.param<double>("~yaw_acc_lim", yaw_acc_lim_, 0.1);
+    pnh_.param<double>("~decel_factor", decel_factor_, 0.9);
+    pnh_.param<int>("~max_count", max_count_, 5);
+    loadParams();
+    pub_vel_ = boost::shared_ptr<geometry_msgs::Twist>(new geometry_msgs::Twist());
+    pub_ = nh_.advertise<geometry_msgs::Twist>("output", 100);
+    sub_ = nh_.subscribe("input", 1, &VelocitySmootherNode::velCallback, this);
+    timer_ = nh_.createTimer(ros::Duration(1./(desired_rate_ * 2.2)), &VelocitySmootherNode::timerCallback, this);
```
Comment (garaemon) [inline cmd_vel_smoother/src/cmd_vel_smoother.cpp]: 2.2?

File: cmd_vel_smoother/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  ${Boost_INCLUDE_DIRS}
+)
+
+add_executable(cmd_vel_smoother src/cmd_vel_smoother.cpp)
+target_link_libraries(cmd_vel_smoother
+  ${catkin_LIBRARIES}
+  ${Boost_LIBRARIES}
+)
+
+install(TARGETS cmd_vel_smoother cmd_vel_smoother
```
Comment (garaemon) [inline cmd_vel_smoother/CMakeLists.txt]: duplicated `cmd_vel_cmoother`

**Comment by garaemon** [comment]
Add README.md and include picture in it.

**Comment by k-okada** [comment]
Hunm this is not `solution`..., just workaround

1) does move_base always publishes unsmoothed cmd_val or only when the load is high?
2) did you checked acceleration value in move_base ?

--
◉ Kei Okada

2015/10/15 0:56、Furushchev <notifications@github.com> のメッセージ:

**Comment by garaemon** [comment]
Is it a LPF? Shumit trigger?

**Comment by garaemon** [comment]
Please add document about algorithm (as equation?)

**Comment by furushchev** [comment]
I think this issue is truly workaround, and need to fix `move_base`.
I created the issue https://github.com/ros-planning/navigation/issues/405 to discuss.
So I close this PR till the end of upstream discussion.

**Comment by k-okada** [comment]
https://github.com/ros-planning/navigation/issues/405#issuecomment-163341269
DLu asking about any updates @furushchev

--
◉ Kei Okada

**Comment by furushchev** [comment]
@k-okada Thank you for reminder. I responded.

**Pull request closed without merging**